### PR TITLE
Implement LLVM-based Lexer for IR

### PR DESCRIPTION
**Pull request by fivosts**
Adds 2 observations:
- Lexedir
- LexedirTuple

### CODE REVIEW

**Comment by fivosts** [comment]
@hughleat

**Comment by fivosts** [comment]
Not sure why 7 commits show up in the PR instead of just 2, probably because I branched from `stable`.

File: compiler_gym/envs/llvm/BUILD [hunk 1]
```diff
@@ -54,13 +54,20 @@ py_library(
     ],
 )
 
+py_library(
+    name = "lexed_ir",
+    srcs = ["lexed_ir.py"],
+    deps = [],
```
Comment (ChrisCummins) [inline compiler_gym/envs/llvm/BUILD]:
```suggestion
```

nitpick

File: compiler_gym/envs/llvm/CMakeLists.txt [hunk 2]
```diff
@@ -56,6 +56,15 @@ cg_py_library(
   PUBLIC
 )
 
+cg_py_library(
+  NAME
+    lexed_ir
+  SRCS
+    "lexed_ir.py"
+  DEPS
```
Comment (ChrisCummins) [inline compiler_gym/envs/llvm/CMakeLists.txt]:
```suggestion
```
nitpick

File: compiler_gym/envs/llvm/service/BUILD [hunk 3]
```diff
@@ -255,6 +255,7 @@ cc_library(
         ":ObservationSpaces",
         "//compiler_gym/service/proto:compiler_gym_service_cc_grpc",
         "//compiler_gym/third_party/autophase:InstCount",
+        "//compiler_gym/third_party/LexedIR:LexedIR",
```
Comment (ChrisCummins) [inline compiler_gym/envs/llvm/service/BUILD]:
```suggestion
        "//compiler_gym/third_party/LexedIR",
```
nitpick

File: compiler_gym/envs/llvm/service/Observation.cc [hunk 4]
```diff
@@ -228,6 +229,38 @@ Status setObservation(LlvmObservationSpace space, const fs::path& workingDirecto
     case LlvmObservationSpace::BUILDTIME: {
       return benchmark.computeBuildtime(reply);
     }
+    case LlvmObservationSpace::LEXEDIR: {
```
Comment (ChrisCummins) [inline compiler_gym/envs/llvm/service/Observation.cc]:
```suggestion
    case LlvmObservationSpace::LEXED_IR: {
```

Forgive my bikeshedding 🙂  LEXED_IR will translate to pascal case "LexedIr" on the python side

File: compiler_gym/envs/llvm/specs.py [hunk 5]
```diff
@@ -38,6 +38,7 @@ class observation_spaces(Enum):
     InstCountNorm = "InstCountNorm"
     InstCountNormDict = "InstCountNormDict"
     AutophaseDict = "AutophaseDict"
+    LexedIR = "LexedIR"
```
Comment (ChrisCummins) [inline compiler_gym/envs/llvm/specs.py]:
```suggestion
    LexedIr = "LexedIr"
```

case-sensitive

File: compiler_gym/third_party/LexedIR/BUILD [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+load("@rules_cc//cc:defs.bzl", "cc_binary", "cc_library")
+
+# py_library(
+#     name = "llvm",
+#     srcs = ["__init__.py"],
+#     visibility = ["//visibility:public"],
+#     deps = [
+#         "//compiler_gym/util",
+#     ],
+# )
```
Comment (ChrisCummins) [inline compiler_gym/third_party/LexedIR/BUILD]: don't commit commented-out code

File: compiler_gym/third_party/LexedIR/decoy_main.cpp [hunk 7]
```diff
@@ -0,0 +1,35 @@
+#include "HughsLexer.h"
```
Comment (ChrisCummins) [inline compiler_gym/third_party/LexedIR/decoy_main.cpp]: Is this file needed?

**Review by ChrisCummins — COMMENTED** [review]
Looking good, thanks @fivosts! I've left some very nitpicky comments inline. Few small things:
- rebase on `development`.
- you should run the pre-commit hooks [here](https://github.com/facebookresearch/CompilerGym/blob/development/.pre-commit-config.yaml#L12) which will sadly obliterate all your nice vertical aligned code 🙂 
- I think we need a copy of the LLVM License file at `compiler_gym/third_party/LexedIR/LICENSE.txt`. You can copy the one from `compiler_gym/third_party/llvm/LICENSE.txt`.
- I don't see any changes to the tests, so `development/tests/llvm/observation_spaces_test.py` should break.  Could you add a test to cover the new space? Take a look [here](https://github.com/facebookresearch/CompilerGym/blob/development/tests/llvm/observation_spaces_test.py#L382-L397) for an example

Cheers,
Chris

File: compiler_gym/third_party/LexedIR/decoy_main.cpp (hunk 7 above)
Comment (fivosts) [reply compiler_gym/third_party/LexedIR/decoy_main.cpp]: Nope, it's a test main that I never removed. Deleting...

File: compiler_gym/third_party/autophase/compute_autophase-prelinked [hunk 8]
```diff
@@ -0,0 +1 @@
+/private/home/foivos/CompilerGym/cmake_build/compiler_gym/third_party/autophase/compute_autophase
```
Comment (ChrisCummins) [inline compiler_gym/third_party/autophase/compute_autophase-prelinked]: Looks like this was checked in by accident

File: compiler_gym/third_party/programl/compute_programl [hunk 9]
```diff
@@ -0,0 +1 @@
+/private/home/foivos/CompilerGym/cmake_build/compiler_gym/third_party/programl/compute_programl
```
Comment (ChrisCummins) [inline compiler_gym/third_party/programl/compute_programl]: Same here

File: tests/llvm/observation_spaces_test.py [hunk 10]
```diff
@@ -62,7 +62,8 @@ def test_observation_spaces(env: LlvmEnv):
         "IrSha1",
         "IsBuildable",
         "IsRunnable",
-        "ObjectTextSizeBytes",
+        "LexedIr",
+        "LexedIrTuple" "ObjectTextSizeBytes",
```
Comment (ChrisCummins) [inline tests/llvm/observation_spaces_test.py]:
```suggestion
        "LexedIrTuple",
        "ObjectTextSizeBytes",
```

typo

**Review by ChrisCummins — COMMENTED** [review]
Thanks for addressing my nitpicks. Here's a couple more 🙂 

Something looks fishy about the git history here (130 commits!), may need rebasing/cherry-picking

Cheers,
Chris

**Comment by ChrisCummins** [comment]
Note the "Pre-commit" job is failing with a few BUILD file errors. You could run buildifier locally or just copy the patch from the CI output

**Comment by fivosts** [comment]
Rebasing to dev probably did this. Did fresh rebase, removed broken symlinks, ran buildifier and cherry picked my commits, should be fine now. Let me know if there's anything else I have missed

**Comment by ChrisCummins** [comment]
LGTM. The CI test failures are known problems and not caused by this change. Merging, thanks @fivosts!

Cheers,
Chris
